volume/drivers/coprhd/coprhd.go

```diff
@@ -186,71 +274,205 @@ func (d *driver) Alerts(volumeID string) (*api.Alerts, error) {
+		return err
+	}
+
+	volume.AttachPath = mountpath
+
+	if err := d.UpdateVol(volume); err != nil {
+		return err
+	}
 	return nil
 }
 
 func (d *driver) Unmount(volumeID string, mountpath string) error {
+	volume, err := d.GetVol(volumeID)
+	if err != nil {
+		return fmt.Errorf("Failed to locate attached volume %q", volumeID)
+	}
+	if err := syscall.Unmount(mountpath, 0); err != nil {
+		return err
+	}
+
+	volume.AttachPath = ""
+
+	if err := d.UpdateVol(volume); err != nil {
+		return err
+	}
 
 	return nil
 }
 
-func (d *driver) Set(
-	volumeID string,
-	locator *api.VolumeLocator,
-	spec *api.VolumeSpec) error {
+func (d *driver) Set(volumeID string, locator *api.VolumeLocator, spec *api.VolumeSpec) error {
 	return volume.ErrNotSupported
 }
 
 func (d *driver) Shutdown() {
 	dlog.Infof("%s Shutting down", Name)
 }
 
-func (d *driver) Snapshot(
-	volumeID string,
-	readonly bool,
-	locator *api.VolumeLocator) (string, error) {
-	return "", nil
+func (d *driver) Snapshot(volumeID string, readonly bool, locator *api.VolumeLocator) (string, error) {
+	return "", volume.ErrNotSupported
 }
 
 func (v *driver) Status() [][2]string {
 	return [][2]string{}
 }
 
-// getAuthSession returns an authenticated API Session
-func (d *driver) getAuthSession() (session *napping.Session, err error) {
-	e := ApiError{}
-
-	s := napping.Session{
-		Userinfo: d.creds,
-		Client:   d.httpClient,
+func (d *driver) Inspect(volumeIDs []string) ([]*api.Volume, error) {
+	volumes, err := d.DefaultEnumerator.Inspect(volumeIDs)
+	if err != nil {
+		return nil, err
 	}
 
-	url := d.url + loginUri
-
-	resp, err := s.Get(url, nil, nil, &e)
+	for _, volume := range volumes {
+		protocol := d.initiator.Protocol
+		device := ""
+
+		switch protocol {
+		case coprhd.InitiatorTypeScaleIO:
+			device, err = d.waitAttachDevice(volume.Id, time.Second*1)
+			if err == nil {
+				volume.DevicePath = device
+			}
+		default:
+		}
+	}
+	return volumes, nil
+}
 
+func (d *driver) Format(volumeID string) error {
+	volume, err := d.GetVol(volumeID)
 	if err != nil {
-		return
+		return fmt.Errorf("Volume is not attached")
 	}
 
-	token := resp.HttpResponse().Header.Get("X-SDS-AUTH-TOKEN")
+	fstype := volume.Spec.Format.SimpleString()
+	devicePath := volume.DevicePath
 
-	h := http.Header{}
-
-	h.Set("X-SDS-AUTH-TOKEN", token)
+	cmd := "/sbin/mkfs." + fstype
+	if _, err := exec.Command(cmd, devicePath).Output(); err != nil {
+		return err
+	}
+	volume.Format = volume.Spec.Format
+	if err := d.UpdateVol(volume); err != nil {
+		return err
+	}
+	return nil
+}
 
-	session = &napping.Session{
-		Client: d.httpClient,
-		Header: &h,
+func (d *driver) waitAttachDevice(volumeID string, to time.Duration) (string, error) {
+	if _, err := d.GetVol(volumeID); err != nil {
+		return "", fmt.Errorf("Volume is not attached")
 	}
 
-	return
+	timeout := time.After(to)
+	timer := time.Tick(time.Millisecond * 100)
+
+	protocol := d.initiator.Protocol
+	device := ""
+
+	for {
+		switch protocol {
```

Comment:
Same as above, leaving the switch for now since it is anticipated that more storage protocols will be request.